mmm_experiments/agents/scientific_value.py

```diff
@@ -0,0 +1,252 @@
+        p = possible_point.reshape(1, -1)
+        D = distance_matrix(p, possible_coordinates).squeeze()
+        argsorted = np.argsort(D)
+        for index in argsorted:
+            if D2[index]:
+                actual_points.append(possible_coordinates[index])
+                break
+
+    return np.array(actual_points)
+
+
+def scientific_value_function(
+    X,
+    Y,
+    sd=None,
+    multiplier=1.0,
+    y_distance_function=None
+):
+    """The value of two datasets, X and Y. Both X and Y must have the same
+    number of rows. The returned result is a value of value for each of the
+    data points.
+
+    Parameters
+    ----------
+    X : numpy.ndarray
+        The input data of shape N x d.
+    Y : numpy.ndarray
+        The output data of shape N x d'. Note that d and d' can be different
+        and they also do not have to be 1.
+    sd : float, optional
+        Controls the length scale decay. We recommend this be set to ``None``
+        to allow for automatic detection of the decay length scale(s).
+    multiplier : float, optional
+        Multiplies the automatically derived length scale if ``sd`` is
+        ``None``.
+    y_distance_function : callable, optional
+        A callable function which takes the array ``Y`` as input and returns
+        an N x N array in which the ith row and jth column is the distance
+        measure between points i and j. Defaults to
+        ``scipy.spatial.distance_matrix`` with its default kwargs (i.e. it is
+        the L2 norm).
+
+    Returns
+    -------
+    array_like
+        The value for each data point.
+    """
+
+    X_dist = distance_matrix(X, X)
+
+    if sd is None:
+        distance = X_dist.copy()
+        distance[distance == 0.0] = np.inf
+        sd = distance.min(axis=1).reshape(1, -1) * multiplier
+
+    # We can make this more pythonic but it makes sense in this case to keep
+    # the default behavior explicit
+    if y_distance_function is None:
+        Y_dist = distance_matrix(Y, Y)
+    else:
+        Y_dist = y_distance_function(Y)
+
+    v = Y_dist * np.exp(-(X_dist**2) / sd**2 / 2.0)
+
+    return v.mean(axis=1)
+
+
+class ScientificValueAgentMixin:
+    """This is a jack-of-all-trades agent which is general to any type of
+    observation (i.e. any beamline). This is a _parameter free_ method, meaning
+    that the user does not need to set any parameters explicitly, though one
+    can manually set a length scale if desired.
+
+    Assumes that ``device``, ``relative_bounds`` and ``measurement_origin`` are
+    defined as attributes.
+    """
+
+    def __init__(
+        self,
+        length_scale: Optional[float] = None,
+        y_distance_function: Optional[callable] = None,
+        optimize_acqf_kwargs: dict = {
+            "q": 1, "num_restarts": 5, "raw_samples": 20
+        },
+        possible_coordinates: Optional[np.ndarray] = None,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self._length_scale = length_scale
```

Comment:
```suggestion
        *,
        device: Union[torch.device, str],
        length_scale: Optional[float] = None,
        y_distance_function: Optional[callable] = None,
        optimize_acqf_kwargs: dict = {
            "q": 1, "num_restarts": 5, "raw_samples": 20
        },
        possible_coordinates: Optional[np.ndarray] = None,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.device = device
        self._length_scale = length_scale
```
It won't always be safe to assume `device`, as not everything is expecting GPU accelerated libraries.